0018-connector-risk-mitigations/0018-connector-risk-mitigations.md

```diff
@@ -17,25 +17,28 @@ The main risk connectors face in Interledger is being unable to fulfill the inco
 Possible mitigations include:
 
 * **Connecting Reliable Ledgers** - Connectors should choose the ledgers they connect in part based on the reliability and speed of fulfillment notifications and the time the ledgers require to process fulfillments once they are submitted
-* **Adjusting Transfer Expiry Window** - Connectors determine the window of time they require between when the incoming and outgoing transfers expire. Connectors should set this window such that they are confident that they will be able to deliver the fulfillment in time even if the outgoing transfer is executed at the last possible moment.
 * **Denial of Service Protections** - Connectors should take standard [Denial of Service](https://en.bitcoin.it/wiki/Weaknesses#Denial_of_Service_.28DoS.29_attacks) attack prevention steps to prevent attackers from overloading the connectors' servers with network traffic
 * **Prioritizing Fulfillments** - Connectors should prioritize fulfilling transfers over preparing new ones or responding to quote requests and may even have separate processes or machines responsible for those different behaviors
 * **Redundant Instances** - Connectors should run redundant processes and machines to increase the difficulty of interfering with their processing
 * **Avoiding Public Endpoints** - Connectors may reduce attack vectors by using communication channels and protocols that do not entail having a publicly-accessible server (for example using [WebRTC](https://webrtc.org/) or [Virtual Private Networks](https://en.wikipedia.org/wiki/Virtual_private_network) to communicate with peers
 
+Note that newer unconditional payment channels in ILPv4 do not require longer timeouts while enabling peers to rebalance frequently and minimize their bilateral trust.
+
 ## Liquidity Exhaustion
 
-Attackers could tie up connectors' liquidity by preparing payments through them that the attacker knows will fail. For example, an attacker could prepare numerous payments to itself and then not fulfill any of them.
+Attackers could tie up connectors' liquidity by preparing payments through them that the attacker knows will fail. For example, an attacker could prepare numerous payments to itself and then not fulfill any of them. In the case of ILPv4, this would require a significant amount of payments to cause damage because of the nature of streaming micropayments.
```

Comment:
> this would require a significant amount of payments to cause damage because of the nature of streaming micropayments

That sounds like a mitigation but it's not really, because it's not too difficult for an attacker to send many packets.

The real mitigation here is that with ILPv4, we adjusted the senders' expectations so that they can function even if the payment bandwidth / liquidity they are allocated is very low. This changed expectation allows the connectors to split their liquidity among their customers such that one client can't interfere with another's connection. (If you're doing large atomic or atomic multi-path payments, you can't limit the amount of liquidity you're allocating to each customer as easily)

Reply:
Also, it seems like an attacker can't really tie up liquidity like they can in HTLCs - unless they are sending multiple prepares from multiple open channels (pretty expensive to do) and the packet amounts are small (the default), technically you can aggressively ban senders that don't complete their prepare-fulfill cycles? Unless I'm missing something.

Reply:
> If you're doing large atomic or atomic multi-path payments

Also, what are "atomic" payments in Interledger? Are these conditional payment channels?

Reply:
> Is this the default or does a connector have to specify it?

Connectors specify the bandwidth they want to allocate to each user (in the `ilp-connector` implementation this is governed by the throughput middleware). Right now on the network the bandwidth is low so it's reasonable for the sender to assume they have pretty low bandwidth.

> technically you can aggressively ban senders that don't complete their prepare-fulfill cycles?

We considered that earlier but decided that isn't the way to go. A) it isn't necessarily the sender's fault if packets are rejected b) rejecting packets may sometimes be useful for application layer protocols c) an attacker could run the same attack where they send money to themselves and do actually fulfill it, but after a delay.

> Also, it seems like an attacker can't really tie up liquidity like they can in HTLCs

If you're not using HTLCs it may be a little more flexible but not completely. Let's say I'm your upstream provider, so you send some big chunk of your packets through me. There's some limit to how much I'll let you send at once (that's bound by our credit limit). If one of your users prepares a ton of packets that go through our connection, they might be able to use up all of your upstream bandwidth, which would block other users from sending.

> Also, what are "atomic" payments in Interledger? Are these conditional payment channels?

There aren't any. I was referring to Lightning's Atomic Multipath Payment (AMP) protocol. We went down the [bulk fulfill](https://github.com/interledger/rfcs/issues/298) path earlier and decided not to implement it.